packages/vite/src/node/utils.ts

```diff
@@ -663,3 +663,26 @@ export function parseRequest(id: string): Record<string, string> | null {
   }
   return Object.fromEntries(new URLSearchParams(search.slice(1)))
 }
+
+// TS rewrite of https://github.com/rxaviers/async-pool/blob/master/lib/es7.js
+// Credits: Rafael Xavier de Souza http://rafael.xavier.blog.br
+export async function asyncPool<Item, Result>({
+  concurrency,
+  items,
+  fn
+}: {
+  concurrency: number
+  items: Item[]
+  fn: (item: Item) => Promise<Result>
+}): Promise<Result[]> {
+  const promises: Promise<Result>[] = []
+  const pool = new Set<Promise<Result>>()
+  for (const item of items) {
+    const promise = fn(item)
+    promises.push(promise)
+    pool.add(promise)
+    promise.then(() => pool.delete(promise))
```

Comment:
shouldn't it be `finally`?
not so critical, as you eventually return promise.all.
but, i think it can stuck the chain at some point, if you have more rejects than the concurrency number.

```suggestion
    promise.finally(() => pool.delete(promise))
```

Reply:
I think the function behave like Promise.all: it throws at the first rejection

Reply:
true, but this extra unhandled promise will trigger an `Uncaught Error` (if rejected, same with finally), so you probably want to catch it to avoid this extra error log.
(the original/linked impl didnt have this issue because the chained promise was put in the pool, and not left hanging)

Reply:
though even the original impl is faulty in some corner case, i.e. when `executing` isn't raced upon:

```
> await asyncPool(2, [0,1,2], (i,a) => i<a.length-1 ? Promise.resolve(i) : Promise.reject(i)).catch(()=>{})
undefined
> Uncaught 2 <-- shouldn't be logged as we catch exceptions
```